src/goto-programs/link_goto_model.cpp

```diff
@@ -46,16 +46,46 @@ static bool link_functions(
   goto_functionst &dest_functions,
   const symbol_tablet &src_symbol_table,
   goto_functionst &src_functions,
-  const rename_symbolt &rename_symbol,
+  rename_symbolt &rename_symbol,
   const std::unordered_set<irep_idt, irep_id_hash> &weak_symbols,
   const replace_symbolt &object_type_updates)
 {
   namespacet ns(dest_symbol_table);
   namespacet src_ns(src_symbol_table);
 
+  // remove any unnecessary duplicates
+  std::unordered_set<irep_idt, irep_id_hash> duplicates;
+  for(const auto &f_pair : src_functions.function_map)
+  {
+    rename_symbolt::expr_mapt::iterator rename_it =
+      rename_symbol.expr_map.find(f_pair.first);
+    if(rename_it == rename_symbol.expr_map.end())
+      continue;
+
+    std::string new_name = id2string(rename_it->second);
+    std::string::size_type suffix_pos = new_name.find("$link");
```

Comment:
Instead of this grungy string decomposition, how about annotating such symbols with a comment `ID_C_original_function_identifier` or such?

Reply:
That would seem nice, but I'm struggling to find a good place to put it in a `symbolt`?

Reply:
I've been sticking annotations like this on the type myself. We should agree on a policy though, or make a breaking change to add `symbolt::attributes`